sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ShowDatabasesExec.scala

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalog.v2.CatalogV2Implicits.NamespaceHelper
+import org.apache.spark.sql.catalog.v2.SupportsNamespaces
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.encoders.RowEncoder
+import org.apache.spark.sql.catalyst.expressions.{Attribute, GenericRowWithSchema}
+import org.apache.spark.sql.catalyst.util.StringUtils
+import org.apache.spark.sql.execution.LeafExecNode
+
+/**
+ * Physical plan node for showing databases.
+ */
+case class ShowDatabasesExec(
+    output: Seq[Attribute],
+    catalog: SupportsNamespaces,
+    pattern: Option[String])
+    extends LeafExecNode {
+  override protected def doExecute(): RDD[InternalRow] = {
+    val namespaces = catalog.listNamespaces().flatMap(getNamespaces(catalog, _))
```

Comment:
This shouldn't list the entire space. It should only call `listNamespaces` once. If the current namespace is and empty array then call `listNamespaces()` and if it is anything else, call `listNamespaces(current)`.

From the [SPIP](https://docs.google.com/document/d/1jEcvomPiTc5GtB9F7d2RTVVpMY64Qy7INCA_rFEd9HQ/edit#heading=h.dwrdqt8xe2wu):

> For a given operation, Spark will call the corresponding catalog method once. For example, SHOW TABLES will return results from listTables(currentNamespace). Spark will not traverse nested namespaces with multiple calls to listNamespaces and listTables.

Reply:
> If the current namespace is and empty array then call listNamespaces()

I just realized that this isn't the same behavior as v1. In v1, `SHOW DATABASES` ignores the current database because databases aren't nested. It always lists all databases (then filters).

The proposed behavior of `SHOW NAMESPACES` was to respect the current namespace and list namespaces nested in it.

There are a few options to fix this:
* Add `SHOW NAMESPACES` that behaves differently than `SHOW DATABASES`
* Make `SHOW NAMESPACES` list all namespaces recursively, like this PR
* Make `SHOW NAMESPACES` list the namespace above the current. If `current=a.b`, then list `a` and show the results (including `b`).
* Change the behavior of `SHOW DATABASES` to match `SHOW NAMESPACES` and list the current
* Change the behavior of `SHOW DATABASES` to match `SHOW NAMESPACES` and list the current, but match behavior if the current namespace is "default"

@imback82, @brkyvz, @cloud-fan, @mccheah, any opinion here?

Reply:
> Add SHOW NAMESPACES that behaves differently than SHOW DATABASES

I prefer this.

Another idea is: `SHOW NAMESPACES` should list the root namespaces of the current catalog, no matter what the current namespace is.

Reply:
Another option is to ignore the current namespace entirely. `SHOW NAMESPACES` would list the root, and `SHOW NAMESPACES IN ns1` lists namespaces in `ns1`. The context is always explicit.

I think I would prefer that option.

Reply:
> Another option is to ignore the current namespace entirely. `SHOW NAMESPACES` would list the root, and `SHOW NAMESPACES IN ns1` lists namespaces in `ns1`. The context is always explicit.

I like this idea. @cloud-fan are you OK with this approach?

Reply:
Yea that's exactly what I mentioned before, +1

> Another idea is: SHOW NAMESPACES should list the root namespaces of the current catalog, no matter what the current namespace is.

Reply:
@cloud-fan @rdblue Thanks for your suggestions.